### DISCUSSION TITLE
Quick Preview Pane - Hotkey to apply focus

### DISCUSSION THREAD

**Discussion by advice2020** [doublecmd/doublecmd#1281] [Suggestions]
I apologize as I am still a little confused how hotkeys are handled by / within WLX plugins in the Quick Preview pane.

Not sure if this is possible.
Allow "tab" key to be able to apply focus to the QPP so can use keyboard page up / down for example when viewing documents.  Also add the ability to be able to hit "tab" again to be able to apply focus back to the other file pane.
Would function exactly how tab works between two file panes.
This way mouse would not be needed for this purpose.

Thank You

**Comment by alexx2000** [comment]
It is disabled because no way to return back.

**Comment by advice2020** [comment]
@Alexx2000
Thank you for your response.
Ahh, I mentioned that on the forum, I figured that was the tricky part to be able to tab back out.  I obviously don't know about the coding aspect, I was just hoping there maybe was some kind of trick that DC would be able to override the plugins focus somehow.  Might be nice to at least be able to tab into the QPP if that is possible, but that is obviously up to.  It would only be worth it if it would allow access to the page up/down & maybe arrows up/down keys, which I am not sure is possible either.
Thank you again for your response.

**Comment by advice2020** [comment]
@alexx2000
DC is very keyboard based so it just feels strange now when certain items have mouse only functionality, so just trying to think of ways to offer a keyboard option for this.  Not sure if others would agree but I think this feature would be nice to have if there is any way possible.
Really feels strange that there is no ability to achieve this.  Anyways, that is the only reason why I am trying hard to come up with ideas for this one.
Again I am obviously not familiar with the coding, so I am sure there are reasons why these things cannot be achieved that I am unaware of.  Obviously you know a thing or two about the program :) so there is no intension other than hoping there is solution for this.
 I am not really expecting a response, but just thought I would share some more observations/questions about this topic.

-----------------------------------------------------------------------------------------------------

So would be able to "tab" key into the QPP but cannot "tab" back out into the other file pane.

1.) Thought someone mentioned this, but doesn't DC have the "hierarchy" as far as keyboard commands / access over plugins?
Wouldn't be able to some how override the plugin and be able to use the "tab" key for DC to jump back into the file pane?

2.) I have noticed that I can access other keyboard hotkeys when the QPP has focus
Ctrl+P allows the QPP to closed

I don't have any hotkey to get into the QPP to give it focus, but I do have a way to jumping out of it with a hotkey.
Found this one by accident when trying to see if a WLX plugin had a find feature, but if you hit Ctrl+F, it activates DC filter hotkey, brings its focus into the filter field, then I hit escape and it brings focus back into the file pane and the QPP remains open in the other pane.
So Ctrl+F works as well

Ctrl+W
Closes tab in other file pane but also closes QPP at same time because tabs is gone

"Enter" key will still open files in associated programs 

-----------------------------------------------------------------------------------------------------

Just strange that the "tab" key would not be able to be used inside plugin, when other hotkey commands could.
Even if it had to be a custom command that users can create, modifier key + tab key.
If can allow the "tab" key to at least enter focus into the QPP then could use the Ctrl+F trick to get out, but not exactly smooth, but it is something.

Again I am just trying to find anything that maybe could spark any ideas.
Anyways, just thought I would share some observations.

Sorry for tagging your name specifically on this one, be this is one I believe only you will know if this is possible or not.
Thank You for taking the time to read this.

**Comment by advice2020** [comment]
Was just wondering if any of this information provided could maybe help with a possible solution for being able to get focus in and out of the preview pane?

**Comment by advice2020** [comment]
Still interested if this could somehow be possible in the future, as mentioned even if it is just being able to use the "Tab" key to enter focus, without a hotkey being able to get out.
Anyways the reason why I am making this reply is that I recently have been using QT based versions of DC and it appears that QT WLX plugins do not allow some of the hotkeys that I mentioned in above posts, such as the Ctrl+Q to close QPP and the Ctrl+F that applies focus to the filter field.  These apparently only worked with the GTK2 version of DC with the GTK2 WLX plugin "gtksocket" when viewing HTML / MHT / PDF files.  Just wanted to clarify after recently discovering this, but it does show that this might be possible in DC.
Would be great if there could be way to allow users the ability to choose focus in and focus out hotkeys.  Not sure if what is happening in DC GTK2 gtksocket plugin can be replicated in DC in general to be used for all plugins, especially QT ones.
Thank You

**Comment by advice2020** [comment]
Probably wasting my time with this but I noticed something I thought I would mention.
I have been playing around with the following QT based WLX plugins
-QMediaPlayer
-QPDFView
-Webengine
This works in all the three plugins mentioned above 
(QMediaPlayer can get focus if you click in the pane above the player controls, pane where video plays)

Say your
-LEFT file pane = your list of files 
-RIGHT file pane = QPP

If your RIGHT file pane contains more than one tab (does NOT work if only one tab)
If you then apply focus into the QPP WLX plugin by clicking your mouse on the QPP
You can use the default DC hotkeys Ctrl+Tab / Ctrl+Shift+Tab and it actually gives the focus back to DC using a keyboard hotkey.  It switches to another tab in the RIGHT file pane, which closes the QPP but the focus is automatically brought back to the LEFT file pane.
Another thing that is strange is that if you create your own custom hotkeys for tab navigation, such as another common set of hotkeys used for tabs, Ctrl+PageUp & Ctrl+PageDown, then this does NOT work.  So it is not the actual command that is allowing this, it has something to do with the actual Ctrl+Tab / Ctrl+Shift+Tab hotkeys themselves, not sure why?

I have not found any tricks to apply focus into the QPP using a keyboard hotkey.
Again though this shows that certain keyboard shortcut keys can be detected and get focus back into DC from the QPP.

Not sure if any of this information gives any abilities / ideas within DC to achieve this.
Just thought I would share what I have found.
Thank You

**Comment by halfhope** [comment]
Hello! 

I’ve added the ability to switch to the quick view window (Ctrl+Q) using Tab, but not back. I’ve also added the ability to show the search dialog by pressing Alt+F and Alt+F3. This code should not be considered stable or fully functional; it’s just a patch.

https://github.com/halfhope/doublecmd/commit/6d156429dc5d55e9b0a9c5ca448c81fb9ce4e153
https://github.com/halfhope/doublecmd/commit/f529dd4ddd818a5480b97c52680a78544be75a3a

**Patch behavior in the built-in viewer**

You can easily navigate within the document in reading mode. Switching back doesn’t work. As far as I understand, implementing backward navigation isn’t difficult, but I couldn’t manage to capture key presses in the image preview window. Instead, I simply close the quick view (Ctrl+Q). I’ve also added the display of the built-in search window using non-conflicting keys: (Dialog) Alt+F / (Next) Alt+F3.

> In the built-in viewer, when focus is transferred, the blinking cursor is not active, which is needed for text navigation. You can enable it in the context menu by clicking on the quick view status bar.

**Patch behavior in WLX plugins**

Focus transfer only works in one direction. ***I could be wrong***, but this depends on window managers like Wayland and Xorg, as they need to pass the window handle of the launched WLX plugin and allow key press interception. As far as I recall, there are no such issues in Windows. Hence, the inability to intercept key presses.

@alexx2000, perhaps it’s worth adding an additional hook to subscribe to keyboard events in the WLX plugin interface? After all, Double Commander is a cross-platform file manager, so it should have its own solutions for different environments.

**Comment by advice2020** [comment]
Thank you for response.
I personally do not really use patches myself.

I agree though, I still think being able to at least "Tab" into the QPP is still valuable.
Allows user to instantly be able to start using keyboard actions/keys in plugins, to navigate through media.

Also I was wondering, what about with the built in internal viewer?
Because it is built in, would it be possible to allow the viewer to receive a `Tab` key to then switch focus back to file pane?
This way if viewing a txt based document, would be able to "Tab" into the document, navigate/read and then "Tab" back to file pane?

Thank You

**Comment by advice2020** [comment]
@alexx2000 

Had a couple questions if interested.
Currently user has to click mouse to get focus in QPP and then also click back into file pane to get out of the QPP = 2 mouse based actions.
Being that the users hands would most likely already be on the keyboard.
Then then would hit TAB key to apply focus into the QPP
Then they would use keyboard navigation keys to navigate inside of the QPP.
Then user would put hand on mouse and click back into other file pane to regain focus....

INTO QPP
Being able to TAB key into the QPP would eliminate 1 of the mouse based actions, which seems like it would still be beneficial.
-You mentioned that there is no problem in getting the focus inside of the QPP.
Would you be willing to allow hotkey to apply focus into the QPP?
-Would it cause any issues / is there a reason to not do this?
Again either way user would have to click out with a mouse, but at least could eliminate 1 mouse action.

-----------------------------------------------------------

OUT OF QPP
You had mentioned that there was no way for DC to get the focus back out of the QPP using hotkey.
-I noticed that some shortcuts are able to be recognized when QPP has focus
For example:
If you have multiple tabs open in file pane and then hit shortcut
CTRL+TAB = Next/Previous Tab = This brings focus back into DC, although it changes to a different tab, which then closes QPP.
But the QPP in this case, is able receive a shortcut from DC and bring focus back into DC.
Couldn't this concept be used to achieve bringing focus back into file pane?
With a setting that would allow user to choose shortcuts for
Focus Into QPP = shortcut
Focus Out Of QPP = shortcut
(also both could be set to the same shortcut if user desires)

-Also DC has comes with 3 WLX plugins/viewers
Image Viewer = does not require focus
Media Player = does not require focus
Text Viewer = would be great if could achieve focus into and out of

Being that the Text Viewer is a built in application (Internal Viewer) within DC, and also does not use the TAB key for anything, if shortcut key into QPP is added, then could shortcut key out of QPP be added for the DC Internal Viewer?
It already is basically possible.
Open `txt` file in QPP
Click mouse into QPP `txt` file to give it focus
Hit F2 key, when this activates `cm_Rename`, the focus is brought back into the other file pane with Inline Rename mode activated.

Would it be considered to add the ability to shortcut key (ideally TAB key by default) out of the QPP `Internal Viewer`?

Thank You

**Comment by advice2020** [comment]
@alexx2000
@rich2014 
Hello again, while I am currently on the subject of the QPP...
I still feel this could be a very helpful update.
I was wondering if anyone would consider at least being able to allow TAB key to apply focus into the QPP.  As mentioned above this could allow user to go from using keyboard in file pane to being able to continue to use keyboard in QPP.  Then could just use mouse to bring focus out of the QPP back into DC.
Would at least eliminate one of the 2 mouse based action that this process currently requires.

Thank You
